conans/client/downloaders/caching_file_downloader.py

```diff
@@ -1,106 +1,165 @@
+        # The recipe path would be different between machines
+        # So best we can do is to set this as unknown
+        summary_key = "unknown"
+
+    if not isinstance(urls, (list, tuple)):
+        urls = [urls]
+    existing_urls = summary["references"].setdefault(summary_key, [])
+    existing_urls.extend(url for url in urls if url not in existing_urls)
+    save(summary_path, json.dumps(summary))
+
+
+def _download_from_urls(urls, file_path, retry, retry_wait, verify_ssl, auth, headers,
+                        md5, sha1, sha256, file_downloader, conanfile):
+    os.makedirs(os.path.dirname(file_path), exist_ok=True)  # filename in subfolder must exist
+    if not isinstance(urls, (list, tuple)):
+        urls = [urls]
+    for url in urls:
+        try:
+            if url.startswith("file:"):
+                file_origin = url2pathname(urlparse(url).path)
+                shutil.copyfile(file_origin, file_path)
+                file_downloader.check_checksum(file_path, md5, sha1, sha256)
             else:
-                ConanOutput()\
-                    .warning("Expected sha256 to be used as file checksums for downloaded sources")
-        if h is None:
-            h = self._get_hash(url, md5, sha1, sha256)
-
-        with self._lock(h):
-            if sources_cache:
-                cached_path = os.path.join(self._download_cache, 's', h)
+                file_downloader.download(url, file_path, retry=retry, retry_wait=retry_wait,
+                                         verify_ssl=verify_ssl, auth=auth, overwrite=True,
+                                         headers=headers, md5=md5, sha1=sha1, sha256=sha256)
+            return
+        except Exception as error:
+            if url != urls[-1]:
+                msg = f"Could not download from the URL {url}: {error}."
+                conanfile.output.warning(msg)
+                conanfile.output.info("Trying another mirror.")
             else:
-                cached_path = os.path.join(self._download_cache, 'c', h)
+                raise
+
+
+class ConanInternalCacheDownloader:
+    """ This is used for the download of Conan packages from server, not for sources
+    """
+    def __init__(self, requester, config):
+        self._download_cache = config.get("core.download:download_cache")
+        if self._download_cache and not os.path.isabs(self._download_cache):
+            raise ConanException("core.download:download_cache must be an absolute path")
+        self._file_downloader = FileDownloader(requester)
+
+    def download(self, url, file_path, auth, verify_ssl, retry, retry_wait):
+        if not self._download_cache:
+            self._file_downloader.download(url, file_path, retry=retry, retry_wait=retry_wait,
+                                           verify_ssl=verify_ssl, auth=auth, overwrite=False)
+            return
+
+        h = compute_sha256(url.encode())
+        cached_path = os.path.join(self._download_cache, "c", h)
```

Comment:
I don't like how to create the class, you now need something that should be specific to the class. But passing it `_download_cache` would also not work (I _think_), so I think will stay as-is :/

Reply:
Changed, with a large refactor to organize it a bit